stacks-signer/src/signerdb.rs

```diff
@@ -1576,6 +1650,66 @@ impl SignerDb {
         ];
         query_row(&self.db, query, args)
     }
+
+    /// Get the received time of the signer state update that achieved a global burn view identified by the provided ConsensusHash
+    pub fn get_burn_block_received_time_from_signers(
+        &self,
+        eval: &GlobalStateEvaluator,
+        ch: &ConsensusHash,
+        local_address: &StacksAddress,
+    ) -> Result<Option<u64>, DBError> {
+        // First make sure we consider the time at which we received the burn block view
+        let local_received_time = self.get_burn_block_receive_time_ch(ch)?;
+
+        let mut vote_weight = 0;
+        if local_received_time.is_some() {
+            vote_weight = eval
+                .address_weights
+                .get(local_address)
+                .copied()
+                .unwrap_or(0);
+        };
+        // Just in case we reached agreement with just our own vote weight...Hopefully this never happens.
+        if eval.reached_agreement(vote_weight) {
+            return Ok(local_received_time);
+        }
+
+        // Then calculate the time at which we received updates from a threshold number of signers with the same burn view
+        let query = r#"
+            SELECT signer_addr, received_time
+            FROM signer_state_machine_updates
+            WHERE burn_block_consensus_hash = ?1
+            ORDER BY received_time ASC
+        "#;
+
+        let mut stmt = self.db.prepare(query)?;
+        let rows = stmt.query_map(params![ch], |row| {
+            let signer_addr: String = row.get(0)?;
+            let received_time: i64 = row.get(1)?;
+            Ok((signer_addr, received_time))
+        })?;
+        for row in rows {
+            let (signer_addr_str, received_time_i64) = row?;
+            let address =
+                StacksAddress::from_string(&signer_addr_str).ok_or(DBError::Corruption)?;
+
+            let weight = eval.address_weights.get(&address).copied().unwrap_or(0);
+            vote_weight = vote_weight.saturating_add(weight);
+
+            if eval.reached_agreement(vote_weight) {
+                let received_time = u64::try_from(received_time_i64).map_err(|e| {
+                    error!("Failed to convert received_time to u64: {e}");
+                    DBError::Corruption
+                })?;
+                return Ok(Some(std::cmp::max(
+                    received_time,
+                    local_received_time.unwrap_or(0),
+                )));
```

Comment:
What is the logic behind using our local received time, if ours is higher than the global state machine?